src/main/java/com/synopsys/integration/alert/channel/azure/boards/actions/AzureBoardsGlobalApiAction.java

```diff
@@ -0,0 +1,101 @@
+        this.configurationAccessor = configurationAccessor;
+        this.descriptorMap = descriptorMap;
+        this.fieldModelConverter = configurationFieldModelConverter;
+    }
+
+    @Override
+    public FieldModel beforeSaveAction(FieldModel fieldModel) throws AlertException {
+        FieldModel updatedFieldModel = super.beforeSaveAction(fieldModel);
+        Optional<DescriptorKey> descriptorKey = descriptorMap.getDescriptorKey(fieldModel.getDescriptorName());
+
+        if (descriptorKey.isPresent()) {
+            ConfigContextEnum context = ConfigContextEnum.valueOf(fieldModel.getContext());
+            List<ConfigurationModel> existingConfig = configurationAccessor.getConfigurationsByDescriptorKeyAndContext(descriptorKey.get(), context);
+            Optional<ConfigurationModel> configurationModel = existingConfig.stream()
+                                                                  .findFirst();
+            updatedFieldModel = updateTokenFields(updatedFieldModel, configurationModel);
+        }
+        return updatedFieldModel;
+    }
+
+    @Override
+    public FieldModel beforeUpdateAction(FieldModel fieldModel) throws AlertException {
+        FieldModel updatedFieldModel = super.beforeUpdateAction(fieldModel);
+        Optional<ConfigurationModel> existingConfig = configurationAccessor.getConfigurationById(Long.valueOf(fieldModel.getId()));
+        return updateTokenFields(updatedFieldModel, existingConfig);
+    }
+
+    private FieldModel updateTokenFields(FieldModel fieldModel, Optional<ConfigurationModel> existingConfig) {
+        Map<String, FieldValueModel> keyToValues = new HashMap<>(fieldModel.getKeyToValues());
+        if (existingConfig.isPresent()) {
+            ConfigurationModel configurationModel = existingConfig.get();
+            Map<String, FieldValueModel> existingFields = fieldModelConverter.convertToFieldValuesMap(configurationModel.getCopyOfFieldList());
+
+            // These fields are saved in the OAuth callback controller so we need to preserve their values on a save or an update.
+            updateMapWithMissingField(AzureBoardsDescriptor.KEY_ACCESS_TOKEN, existingFields, keyToValues);
+            updateMapWithMissingField(AzureBoardsDescriptor.KEY_REFRESH_TOKEN, existingFields, keyToValues);
+            updateMapWithMissingField(AzureBoardsDescriptor.KEY_TOKEN_EXPIRATION_MILLIS, existingFields, keyToValues);
+            updateMapWithMissingField(AzureBoardsDescriptor.KEY_OAUTH_USER_EMAIL, existingFields, keyToValues);
+        }
+
+        return new FieldModel(fieldModel.getDescriptorName(), fieldModel.getContext(), fieldModel.getCreatedAt(), fieldModel.getLastUpdated(), keyToValues);
+    }
+
+    private void updateMapWithMissingField(String key, Map<String, FieldValueModel> databaseFields, Map<String, FieldValueModel> fieldModelFields) {
+        if (databaseFields.containsKey(key)) {
+            fieldModelFields.put(key, databaseFields.get(key));
+        }
```

Comment:
Do we always want to prefer what's in the DB? Or should we use the value of the FieldValueModel if it has one?

Reply:
The fieldModel on a save will not have the token fields.  They aren't rendered at all in the UI, therefore the FieldModel we are trying to save will be missing those values always.